### ISSUE TITLE
Support conditional return types

### ISSUE THREAD

**Issue by muglug** [vimeo/psalm#3005]
TypeScript's [conditional types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#conditional-types) suit the behaviour of a lot of legacy code.

Proposed syntax:

`(<template param> is <union type> ? <union type> : <union type>)`

Enforcing brackets makes parsing for this a lot easier.

In use:

```php
class A {
    /** @var array<string, string> */
    private array $itemAttr = [];
    
    /**
     * @template T as ?string
     * @param T $name
     * @return string|string[]
     * @psalm-return (T is string ? string : array<string, string>)
     */
    public function getAttribute(?string $name, string $default = '')
    {
       if (null === $name) {
            return $this->itemAttr;
        }
        return isset($this->itemAttr[$name]) ? $this->itemAttr[$name] : $default;
    }
}

(new A)->getAttribute("colour", "red"); // typed as string
(new A)->getAttribute(null); // typed as array<string, string>
(new A)->getAttribute($_GET["foo"]); // typed as string|array<string, string>
```

**Comment by weirdan** [comment]
Could partially address #698

**Comment by muglug** [comment]
cc @Ocramius, @TysonAndre and others who might have smart thoughts.

**Comment by muglug** [comment]
Alternatively we could use the keyword `is` (which Hack now uses for runtime type checks):

`(T is string ? string : array<string, string>)`

**Comment by TysonAndre** [comment]
Pattern matching was another idea I had for syntax for use cases like this. That might also be useful for user-defined functions/methods that act like var_export.

This would let you make types depend on a combination of parameters, and avoids needing templates except when necessary.

There's also the question of what to do if the caller passes through a `?string` - I assume both proposals would infer `string|array<string, string>`, and have equivalent functionality with different syntax.

```php
/**
 * @param mixed $expression
 * @psalm-return-matching (false $return) void
 * @psalm-return-matching (true $return) string
 */
function var_export($expression, bool $return = false) {
    // ...
}
```

```php
    /**
     * @return string|string[]
     * @psalm-return-matching (string $name) string
     * @psalm-return-matching (null $name) array<string, string>
     */
    public function getAttribute(?string $name, string $default = '')
    {
       if (null === $name) {
            return $this->itemAttr;
        }
        return isset($this->itemAttr[$name]) ? $this->itemAttr[$name] : $default;
    }
```

**Comment by muglug** [comment]
> There's also the question of what to do if the caller passes through a ?string - I assume both proposals would infer string|array<string, string>, and have equivalent functionality with different syntax.

Yeah

> Pattern matching was another idea I had for syntax for use cases like this. That might also be useful for user-defined functions/methods that act like var_export.

The reason I prefer the originally-proposed syntax is that, even though it's a little more verbose for the end user (the `@template` params have to be given), it fits much more neatly into Psalm's existing mechanics – beyond the type parsing, there's not _too_ much extra that needs to be done.

**Comment by TysonAndre** [comment]
Is the `T` necessary, though? Could it just be

`     * @psalm-return ($x extends string ? string : array<string, string>)`

**Comment by muglug** [comment]
Even if Psalm did support that syntax, it would still use templates under the hood because templates are well-understood – they're bound at call-time, so calculating the resultant type is pretty trivial.

I also find it useful, when venturing into foreign waters, to stay fairly close to a big ship – TypeScript, in this instance, which uses `T`.

**Comment by Ocramius** [comment]
How about using substitution with `callable`?

```
@psalm-match callable (T1, T2=): T3
@psalm-match callable (T4, T5): T6
function foo($a, $b) {
    // ...
}
```

That means that the `@psalm-match` takes priority (when matched) over the type of `foo` that is inferred from its signature.

Effectively, it's method overloading.

**Comment by Ocramius** [comment]
```php
/**
 * @param mixed $expression
 * @psalm-match callable(mixed, true) : string
 * @psalm-match callable(mixed, bool=) : void
 */
function var_export($expression, bool $return = false) {
    // ...
}
```

**Comment by Ocramius** [comment]
Also interesting:

```php
/**
 * @param int|float $a
 * @param int|float $b
 * @return int|float
 * @psalm-match callable(float, int) : float
 * @psalm-match callable(int, float) : float
 * @psalm-match callable(int, int) : int
 * @psalm-match callable(float float) : float
 */
function add($a, $b) {
    return $a + $b;
}
```

It can also be interesting to verify totality of implementation 👍

**Comment by muglug** [comment]
@Ocramius In the proposed syntax those would become

```php
/**
 * @template TReturnFlag as bool
 * @param mixed $expression
 * @param TReturnFlag $return
 * @psalm-return (TReturnFlag is true ? string : void)
 */
function var_export($expression, bool $return = false) {
    // ...
}
```
and
```
/**
 * @template T as int|float
 * @param T $a
 * @param T $b
 * @return int|float
 * @psalm-return (T is int ? int : float)
 */
function add($a, $b) {
    return $a + $b;
}
```

I think those are both a little easier to understand

**Comment by muglug** [comment]
Current working branch: https://github.com/vimeo/psalm/compare/conditional-types

**Comment by Ocramius** [comment]
Dunno, I think the syntax in https://github.com/vimeo/psalm/issues/3005#issuecomment-602143042 is extremely hard to follow.

Isn't it much easier to match multiple manually declared signatures?

**Comment by muglug** [comment]
> Isn't it much easier to match multiple manually declared signatures?

No – I've just committed a prototype of this to master, and the matching logic for ternary conditional types is (currently at least) very self-contained: https://github.com/vimeo/psalm/commit/6058725256c24841c3d5dad4474696ded73a9968#diff-88aa0097e81fcbafe2955bf5efd0ae34R1270-R1313

Please kick the tyres on this – I'm going to keep this ticket open while I work out kinks.

Here's a working example: https://psalm.dev/r/740dcfbf81

**Comment by Ocramius** [comment]
I don't mind the other syntax if it gets the ball rolling: how does it behave with 3 possible input combinations though? Thinking horrible things like methods that accept bitmasks.

**Comment by muglug** [comment]
This should work (it doesn't right now, but it will):

```
/**
 * @template T as int
 * @param T $i
 * @psalm-return (T is 0 ? string : (T is 1 ? int : bool))
 */
function getDifferentType(int $i) {
    if ($i === 0) {
        return "hello";
    }

    if ($i === 1) {
        return 5;
    }

    return true;
}
```

**Comment by Ocramius** [comment]
If that works, OK: still a mess though 😅

**Comment by muglug** [comment]
Yeah, I guess I'm slightly biased in favour of messiness – I don't want this annotation to become commonplace, because it's antagonistic towards static analysis in general – any function that returns different sorts of things based on explicitly-provided flags should really be a collection of separate functions.

As far as I can see TypeScript seems to support this mainly for legacy use-cases, too.

**Comment by muglug** [comment]
@TysonAndre support for your suggestion – `($x is string : Foo : Bar)` – was added in 56cc5fb.

https://psalm.dev/r/056b0b09ac
